### PR TITLE
Feat(validator-cli)/arb to gnosis

### PR DESCRIPTION
**Pull request by mani99brar**
## PR-Codex overview
This PR primarily focuses on refactoring and enhancing the transaction handling logic for bridging between Ethereum and Arbitrum, introducing better structure, error handling, and new interfaces for transaction handlers.

### Detailed summary
- Deleted obsolete files related to previous transaction handling.
- Updated `getBlockFromEpoch` to fetch final blocks instead of the latest.
- Modified environment variables for subgraph endpoints to include specific chains.
- Refactored `getBotPath` function to return structured config objects.
- Introduced new transaction handler interfaces and improved error handling.
- Added new transaction handlers for `ArbToEth` and `ArbToGnosis`.
- Enhanced tests for transaction handlers to cover new functionalities and edge cases.
- Updated snapshot handling to include chain ID in relevant functions.
- Improved claim handling functions to accept chain ID as a parameter.

> The following files were skipped due to too many changes: `validator-cli/src/utils/transactionHandlers/baseTransactionHandler.ts`, `validator-cli/src/utils/transactionHandlers/baseTransactionHandler.test.ts`

> ✨ Ask PR-Codex anything about this PR by commenting with `/codex {your question}`

## Summary by CodeRabbit

- **New Features**
  - Introduced a modular, extensible transaction handler framework for cross-chain claim and snapshot workflows, supporting both Arbitrum-Ethereum and Arbitrum-Gnosis bridges, including devnet variants.
  - Added support for chain-aware operations and configuration, including multi-network subgraph queries, router RPC provider integration, and dynamic environment variable handling.
  - Enabled new bridge configuration options, including deposit tokens and per-network deposit amounts.
  - Added comprehensive claim challenge and resolution workflows with event-driven status tracking.

- **Bug Fixes**
  - Improved error handling and robustness in snapshot and claim management to prevent duplicate or expired transactions.
  - Added guards to prevent errors from invalid inputs and missing data in snapshot and claim processing.

- **Refactor**
  - Generalized transaction handler logic and interfaces for easier extension and maintenance.
  - Centralized claim, challenge, and snapshot logic into dedicated helper modules.
  - Updated environment variable naming for clarity and network specificity.
  - Replaced direct imports of concrete transaction handlers with interface-based abstractions and dynamic handler retrieval.

- **Tests**
  - Added comprehensive unit tests for base and specific transaction handlers, ensuring reliability of claim, challenge, and snapshot operations.
  - Updated existing tests to accommodate new parameters and refactored interfaces.

- **Chores**
  - Removed obsolete files and legacy logic, streamlining the codebase for maintainability.
  - Updated block fetching to use finalized block tags for improved accuracy.

### CODE REVIEW

File: validator-cli/.env.dist [hunk 1]
```diff
@@ -15,6 +15,9 @@ GNOSIS_AMB_ADDRESS=0x8448E15d0e706C0298dECA99F0b4744030e59d7d
 VEAOUTBOX_CHAINS=11155111,421611
 
 # vescan subgraph endpoints
-VEAINBOX_SUBGRAPH=https://api.studio.thegraph.com/query/user/inbox-arb-sep/version/latest
-VEAOUTBOX_SUBGRAPH=https://api.studio.thegraph.com/query/user/outbox-arb-sep/version/latest
+VEAINBOX_SUBGRAPH_ARBSEPOLIA=https://api.studio.thegraph.com/query/user/inbox-arb-sep/version/latest
+VEAOUTBOX_SUBGRAPH_SEPOLIA=https://api.studio.thegraph.com/query/user/outbox-arb-sep/version/latest
+VEAOUTBOX_SUBGRAPH_CHIADO=https://api.studio.thegraph.com/query/user/outbox-arb-sep/version/latest
```
[the remark this answers is not in the document]
Comment (mani99brar) [reply validator-cli/.env.dist:20]: Its an example file, the actual values will be used .env

File: validator-cli/src/helpers/claimer.ts [hunk 2]
```diff
@@ -60,18 +60,17 @@ async function checkAndClaim({
...
+      transactionHandler as IDevnetTransactionHandler,
       veaInbox,
       emitter
     );
   } else if (claim == null && epoch == claimAbleEpoch) {
-    return makeClaim(epoch, transactionHandler, outboxStateRoot, veaInbox, veaOutbox, fetchLatestClaimedEpoch);
+    return makeClaim(chainId, epoch, transactionHandler, outboxStateRoot, veaInbox, veaOutbox, fetchLatestClaimedEpoch);
   } else if (claim != null) {
     return verifyClaim(transactionHandler, claim, veaOutboxProvider);
   } else {
```
[the remark this answers is not in the document]
Comment (mani99brar) [reply validator-cli/src/helpers/claimer.ts:75]: outboxStateRoot won't change for an epoch we are claiming.

File: validator-cli/src/helpers/validator.ts [hunk 3]
```diff
@@ -61,13 +68,16 @@ export async function challengeAndResolveClaim({
   }
   const ethBlockTag = finalityIssueFlagEth ? "finalized" : "latest";
   if (!transactionHandler) {
-    transactionHandler = new ArbToEthTransactionHandler({
+    const TransactionHandler = fetchTransactionHandler(chainId, Network.TESTNET);
+    transactionHandler = new TransactionHandler({
+      chainId,
       network: Network.TESTNET, // Hardcoded as TESTNET & MAINNET have same contracts
```
[the remark this answers is not in the document]
Comment (mani99brar) [reply validator-cli/src/helpers/validator.ts]: The challenger is only used for Network.TESTNET for now as there is no MAINNET support yet.

File: validator-cli/src/utils/graphQueries.ts [hunk 4]
```diff
@@ -16,14 +16,28 @@
...
+  } else if (chainId === 10200) {
+    return process.env.VEAOUTBOX_SUBGRAPH_CHIADO || "";
+  }
+};
+const getInboxSubgraphUrl = (chainId: number): string => {
+  // using destination chainId's for inbox
+  if (chainId === 11155111 || chainId === 10200) {
+    return process.env.VEAINBOX_SUBGRAPH_ARBSEPOLIA || "";
+  }
+};
```
[the remark this answers is not in the document]
Comment (mani99brar) [reply validator-cli/src/utils/graphQueries.ts:31]: 👍 Will add chain validation in botConfig file.

File: validator-cli/src/utils/graphQueries.ts [hunk 5]
```diff
@@ -136,15 +150,18 @@ type SenSnapshotResponse = {
  * @param epoch
  * @returns snapshot data
  */
-const getSnapshotSentForEpoch = async (epoch: number, veaInbox: any): Promise<{ txHash: string }> => {
+const getSnapshotSentForEpoch = async (
+  epoch: number,
+  veaInbox: string,
+  chainId: number
+): Promise<{ txHash: string }> => {
```
[the remark this answers is not in the document]
Comment (mani99brar) [reply validator-cli/src/utils/graphQueries.ts:157]: 👍 Will do

File: validator-cli/src/utils/transactionHandlers/arbToEthHandler.test.ts [hunk 6]
```diff
@@ -0,0 +1,218 @@
...
+      veaInboxProvider,
+      veaOutboxProvider,
+      emitter: mockEmitter,
+      claim: null,
+    };
+  });
+  describe("makeClaim", () => {
+    let transactionHandler: ArbToEthTransactionHandler;
+    const { routeConfig } = getBridgeConfig(chainId);
+    const deposit = routeConfig[Network.DEVNET].deposit;
```
[the remark this answers is not in the document]
Comment (mani99brar) [reply validator-cli/src/utils/transactionHandlers/arbToEthHandler.test.ts]: 👍

File: validator-cli/src/utils/transactionHandlers/arbToEthHandler.test.ts [hunk 7]
```diff
@@ -0,0 +1,218 @@
...
+      expect(transactionHandler.transactions.executeSnapshotTxn).toEqual({
+        hash: "0x1234",
+        broadcastedTimestamp: expect.any(Number),
+      });
+    });
+
+    it("should not resolve challenged claim if txn is pending", async () => {
+      jest.spyOn(transactionHandler, "checkTransactionStatus").mockResolvedValue(1);
+      transactionHandler.transactions.executeSnapshotTxn = { hash: "0x1234", broadcastedTimestamp: 1000 };
+      await transactionHandler.resolveChallengedClaim(mockMessageExecutor);
```
[the remark this answers is not in the document]
Comment (mani99brar) [reply validator-cli/src/utils/transactionHandlers/arbToEthHandler.test.ts]: 👍

File: validator-cli/src/utils/transactionHandlers/arbToEthHandler.ts [hunk 8]
```diff
@@ -0,0 +1,156 @@
...
+      "challenge(uint256,(bytes32,address,uint32,uint32,uint32,uint8,address))"
+    ].estimateGas(this.epoch, this.claim, { value: deposit });
+
+    // Profit-driven fee calculation
+    const maxFeePerGas = deposit / (gasEstimate * BigInt(6));
+    let maxPriorityFeePerGas = BigInt(6_667_000_000_000);
+    if (maxPriorityFeePerGas > maxFeePerGas) {
+      maxPriorityFeePerGas = maxFeePerGas;
+    }
+
```
[the remark this answers is not in the document]
Comment (mani99brar) [reply validator-cli/src/utils/transactionHandlers/arbToEthHandler.ts]: Yeah correct but what do you mean by illegal?

File: validator-cli/src/utils/transactionHandlers/arbToEthHandler.ts [hunk 9]
```diff
@@ -0,0 +1,156 @@
...
+export interface DevnetTransactions extends Transactions {
+  devnetAdvanceStateTxn: Transaction | null;
+}
+
+export class ArbToEthDevnetTransactionHandler extends ArbToEthTransactionHandler {
+  public veaOutboxDevnet: VeaOutboxArbToEthDevnet;
+  public transactions: DevnetTransactions = {
+    ...this.transactions,
+    devnetAdvanceStateTxn: null,
+  };
```
[the remark this answers is not in the document]
Comment (mani99brar) [reply validator-cli/src/utils/transactionHandlers/arbToEthHandler.ts:132]: We expect all transaction on BaseTransactionHandler to null so not an issue.

File: validator-cli/src/utils/transactionHandlers/arbToGnosisHandler.test.ts [hunk 10]
```diff
@@ -0,0 +1,213 @@
...
+        veaOutbox: veaOutbox,
+        veaInboxProvider: inboxProvider,
+        veaOutboxProvider: outboxProvider,
+        emitter: mockEmitter,
+        claim,
+      });
+    });
+
+    it("throws if claim not set", async () => {
+      const h = new ArbToGnosisTransactionHandler({ ...transactionHandler, claim: null });
```
[the remark this answers is not in the document]
Comment (mani99brar) [reply validator-cli/src/utils/transactionHandlers/arbToGnosisHandler.test.ts]: 👍

File: validator-cli/src/utils/transactionHandlers/arbToGnosisHandler.test.ts [hunk 11]
```diff
@@ -0,0 +1,213 @@
...
+      jest.spyOn(transactionHandler, "checkTransactionStatus").mockResolvedValue(TransactionStatus.NOT_MADE);
+      (messageExecutor as jest.Mock).mockResolvedValue({ hash: "0xexec" });
+    });
+
+    it("throws if claim not set", async () => {
+      await expect(transactionHandler.resolveChallengedClaim("0xtx")).rejects.toThrow(ClaimNotSetError);
+    });
+
+    it("executes and records transaction", async () => {
+      transactionHandler = new ArbToGnosisTransactionHandler({ ...transactionHandler, claim });
```
[the remark this answers is not in the document]
Comment (mani99brar) [reply validator-cli/src/utils/transactionHandlers/arbToGnosisHandler.test.ts]: 👍

File: validator-cli/src/utils/transactionHandlers/arbToGnosisHandler.ts [hunk 12]
```diff
@@ -0,0 +1,148 @@
...
+    const signer = getWallet(privateKey, outboxRPC);
+
+    const weth = getWETH(depositToken, privateKey, outboxRPC);
+    const currentAllowance: bigint = await weth.allowance(signer.address, veaOutbox.address);
+    if (currentAllowance < deposit) {
+      const approvalAmount = deposit * BigInt(1); // Approving for 10 claims
+      const approveTx = await weth.approve(routeConfig[Network.TESTNET].veaOutbox.address, deposit * approvalAmount);
+      await approveTx.wait();
+    }
+  }
```
[the remark this answers is not in the document]
Comment (mani99brar) [reply validator-cli/src/utils/transactionHandlers/arbToGnosisHandler.ts]: Yeah fixed to 10 claims 👍

File: validator-cli/src/utils/transactionHandlers/arbToGnosisHandler.ts [hunk 13]
```diff
@@ -0,0 +1,148 @@
...
+    let maxPriorityFeePerGasMEV = BigInt(6667000000000); // 6667 gwei
+    // Ensure maxPriorityFeePerGas <= maxFeePerGas
+    if (maxPriorityFeePerGasMEV > maxFeePerGasProfitable) {
+      maxPriorityFeePerGasMEV = maxFeePerGasProfitable;
+    }
+    const tx = await this.veaOutbox.challenge(this.epoch, this.claim, {
+      maxFeePerGas: maxFeePerGasProfitable,
+      maxPriorityFeePerGas: maxPriorityFeePerGasMEV,
+      gasLimit: gasEstimate,
+    });
```
[the remark this answers is not in the document]
Comment (mani99brar) [reply validator-cli/src/utils/transactionHandlers/arbToGnosisHandler.ts]: 👍

File: validator-cli/src/utils/transactionHandlers/baseTransactionHandler.test.ts [hunk 14]
```diff
@@ -0,0 +1,259 @@
...
+      await transactionHandler.saveSnapshot();
+      expect(veaInbox.saveSnapshot).not.toHaveBeenCalled();
+    });
+  });
+
+  describe("startVerification()", () => {
+    const cfg = getBridgeConfig(chainId);
+    const flipTime = Number(claim.timestampClaimed) + cfg.sequencerDelayLimit + cfg.routeConfig[network].epochPeriod;
+    it("throws if claim not set", async () => {
+      transactionHandler = new DummyHandler({ ...transactionHandlerParams });
```
[the remark this answers is not in the document]
Comment (mani99brar) [reply validator-cli/src/utils/transactionHandlers/baseTransactionHandler.test.ts:177]: 👍

File: validator-cli/src/utils/transactionHandlers/baseTransactionHandler.test.ts [hunk 15]
```diff
@@ -0,0 +1,259 @@
...
+      expect(transactionHandler.transactions.startVerificationTxn).toBeNull();
+    });
+  });
+
+  describe("verifySnapshot()", () => {
+    const cfg = getBridgeConfig(chainId);
+    const flipTime = Number(claim.timestampVerification) + cfg.minChallengePeriod;
+
+    it("throws if claim not set", async () => {
+      const h = new DummyHandler({ ...transactionHandler, claim: null });
```
[the remark this answers is not in the document]
Comment (mani99brar) [reply validator-cli/src/utils/transactionHandlers/baseTransactionHandler.test.ts]: 👍

File: validator-cli/src/utils/transactionHandlers/baseTransactionHandler.test.ts [hunk 16]
```diff
@@ -0,0 +1,259 @@
...
+      jest.spyOn(transactionHandler, "checkTransactionStatus").mockResolvedValue(TransactionStatus.NOT_MADE);
+      await transactionHandler.withdrawClaimDeposit();
+      expect(veaOutbox.withdrawClaimDeposit).toHaveBeenCalledWith(epoch, claim);
+      expect(transactionHandler.transactions.withdrawClaimDepositTxn).toHaveProperty("hash", "0xc");
+    });
+  });
+
+  describe("withdrawChallengeDeposit()", () => {
+    it("throws if claim not set", async () => {
+      const h = new DummyHandler({ ...transactionHandler, claim: null });
```
[the remark this answers is not in the document]
Comment (mani99brar) [reply validator-cli/src/utils/transactionHandlers/baseTransactionHandler.test.ts]: 👍

File: validator-cli/src/utils/transactionHandlers/baseTransactionHandler.ts [hunk 17]
```diff
@@ -0,0 +1,304 @@
...
+      case ContractType.INBOX:
+        provider = this.veaInboxProvider;
+        break;
+      case ContractType.OUTBOX:
+        provider = this.veaOutboxProvider;
+        break;
+      case ContractType.ROUTER:
+        provider = this.veaRouterProvider;
+        break;
+    }
```
[the remark this answers is not in the document]
Comment (mani99brar) [reply validator-cli/src/utils/transactionHandlers/baseTransactionHandler.ts:167]: Let it crash as this branch should never be reached.

File: validator-cli/src/utils/transactionHandlers/arbToGnosisHandler.ts [hunk 18]
```diff
@@ -0,0 +1,148 @@
...
+    const { veaOutbox, deposit } = routeConfig[this.network];
+    const privateKey = process.env.PRIVATE_KEY!;
+    const signer = getWallet(privateKey, outboxRPC);
+
+    const weth = getWETH(depositToken, privateKey, outboxRPC);
+    const currentAllowance: bigint = await weth.allowance(signer.address, veaOutbox.address);
+    if (currentAllowance < deposit) {
+      const approvalAmount = deposit * BigInt(10); // Approving for 10 claims
+      const approveTx = await weth.approve(routeConfig[Network.TESTNET].veaOutbox.address, deposit * approvalAmount);
+      await approveTx.wait();
```
[the remark this answers is not in the document]
Comment (jaybuidl) [reply validator-cli/src/utils/transactionHandlers/arbToGnosisHandler.ts:37]: which deposit bug?

**Review by jaybuidl — APPROVED** [review]

File: validator-cli/src/utils/transactionHandlers/arbToGnosisHandler.ts (hunk 18 above)
[the remark this answers is not in the document]
Comment (mani99brar) [reply validator-cli/src/utils/transactionHandlers/arbToGnosisHandler.ts:37]: Yes fixing in the next PR.

File: validator-cli/src/utils/graphQueries.ts (hunk 4 above)
[the remark this answers is not in the document]
Comment (mani99brar) [reply validator-cli/src/utils/graphQueries.ts:31]: Adding this is next PR.
